### PR TITLE
Add Dalli instrumentation gem

### PR DESCRIPTION
**Pull request by robertlaurin**
Closes https://github.com/open-telemetry/opentelemetry-ruby/issues/244

### CODE REVIEW

File: instrumentation/dalli/lib/opentelemetry/instrumentation/dalli/utils.rb
Comment (robertlaurin) [inline instrumentation/dalli/lib/opentelemetry/instrumentation/dalli/utils.rb:10]: The utils here are shameless ported from some existing instrumentation, there is some room for feedback/improvement here.

Comment (robertlaurin) [reply instrumentation/dalli/lib/opentelemetry/instrumentation/dalli/utils.rb:10]: https://github.com/open-telemetry/opentelemetry-ruby/issues/339 this is also feels very relevant here

File: instrumentation/dalli/lib/opentelemetry/instrumentation/dalli/patches/client.rb
Comment (ericmustin) [inline instrumentation/dalli/lib/opentelemetry/instrumentation/dalli/patches/client.rb:12]: Haven't used Dalli in production before, could you explain a bit the benefit of instrumenting the higher level `Dalli::Client.perform` and `Dalli::Client.get_multi_yielder`, as opposed to `Dalli::Server.request`?

Basically, by choosing to wrap `Dalli::Client.get_multi_yielder` we introduce a bit of a tradeoff it seems, where we're losing some visibility into the underlying `host/port` info of the server, since `get_multi_yielder` calls `server.request` directly without going thru perform, if i'm reading this correctly? https://github.com/petergoldstein/dalli/blob/ca8b286de230546c3bfe2789e9b6bc5a2c619570/lib/dalli/client.rb#L429

So in those cases we'd lose the  `net.peer.name` and `net.peer.port` attributes, (since capturing it for a query that hits multiple servers introduces a bit of a race condition), and if there are errors we may not be able to attribute it to a specific server. But, as a tradeoff, we get better visibility into the timing of the entire operation, which is what the user is performing and likely cares more about. Am I stating that correctly? And I guess, more importantly is this intentional and something you feel is a greater value add? I defer to your opinion here but think it's worth being explicit about.

My other thought here is maybe we wrap both? IE, wrap the multi yielder, but also generate child spans for each `server.request`. That way if say, there's a connection issue or latency on a specific server, it's clear which one, while with the `get_multi` parent span we also _get_ (_hehe_) the visibility into what the broader user query was? Not sure if that's worth the implementation effort but just throwing it out there

Comment (fbogsany) [reply instrumentation/dalli/lib/opentelemetry/instrumentation/dalli/patches/client.rb:12]: Thanks for the suggestion. When writing our instrumentation, we saw `Client.perform` as the "Chokepoint method for instrumentation", and I guess we didn't dig deeper. Looking at `Server.request`, I see that's also marked as "Chokepoint method for instrumentation", and it is likely better to instrument at that level, closer to the wire. I think we should instrument *only* at that level, rather than introducing additional spans at the higher level.

File: instrumentation/dalli/test/opentelemetry/instrumentation/dalli/instrumentation_test.rb
Comment (fbogsany) [inline instrumentation/dalli/test/opentelemetry/instrumentation/dalli/instrumentation_test.rb:51]: 🤔 I'd prefer to see actual `memcache` commands here. IIRC, this should be a `GET`. For cases where Dalli maps the command names to something else, it'd be good to map them back to the real command names.

File: instrumentation/dalli/lib/opentelemetry/instrumentation/dalli/utils.rb
Comment (ericmustin) [reply instrumentation/dalli/lib/opentelemetry/instrumentation/dalli/utils.rb:10]: Seems fine to me, i don't have strong opinions on the default placeholder but that's the only thing maybe we may want to change. not a big deal at all, ship it imo

**Review by ericmustin — COMMENTED** [review]
Generally this looks good to me, the changes over to `::Server` I think simplify things. All that's left here is probably a test for an error case, and perhaps some tests for the behavior around multi_get/sets. the attributes look in line with the spec, if there's any other optional attributes on there that are reasonable to add, go for it, but if not no worries https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/database.md

Ping me when this is out of draft mode and happy to approve after giving another once over for the above suggestions

Also, great work on this!

File: instrumentation/dalli/test/opentelemetry/instrumentation/dalli/instrumentation_test.rb
Comment (robertlaurin) [inline instrumentation/dalli/test/opentelemetry/instrumentation/dalli/instrumentation_test.rb:76]: I'm calling `.set` so the client sets up a server in the ring instance variable, then grabbing it for a stub. I want to try and simulate an actual error that could crop up.  There wasn't a better way obvious to me on how to do this.

File: instrumentation/dalli/lib/opentelemetry/instrumentation/dalli/utils.rb
Comment (robertlaurin) [inline instrumentation/dalli/lib/opentelemetry/instrumentation/dalli/utils.rb:39]: There are two underlying write requests for the `sasl_authentication` so I'm not sure if it should map to one or the other here.

File: instrumentation/dalli/lib/opentelemetry/instrumentation/dalli/utils.rb
Comment (robertlaurin) [inline instrumentation/dalli/lib/opentelemetry/instrumentation/dalli/utils.rb:43]: I don't like [repeating the multi check](https://github.com/petergoldstein/dalli/blob/master/lib/dalli/server.rb#L289) for getting the op code, but I didn't think of a better way of doing it.  Open to suggestions for improvement here.

File: instrumentation/dalli/test/opentelemetry/instrumentation/dalli/instrumentation_test.rb
Comment (ericmustin) [reply instrumentation/dalli/test/opentelemetry/instrumentation/dalli/instrumentation_test.rb:76]: lgtm, i also have no ideas here off the top of my head

File: instrumentation/dalli/lib/opentelemetry/instrumentation/dalli/utils.rb
Comment (ericmustin) [reply instrumentation/dalli/lib/opentelemetry/instrumentation/dalli/utils.rb:39]: maybe we just add a heuristic in the `opname` method to make this choice, and return like `'auth'` from opname mapping? my thought is it's better to be accurate than necessarily have the tightest code
```
if (temp_opname = OPNAME_MAPPING[lookup_name] || operation.to_s) && temp_opname == 'auth'
  // additional heuristic to check if its negotiation or request
else
  return temp_opname
end
```

**Review by ericmustin — APPROVED** [review]
Left some small feedbacks on remaining open questions but generally comfortable having this merged so happy to approve now, assuming we can clear up those Jruby test failures if they are not transient (tough to tell if they're transient or not, we can discuss at SIG mtg if it's recurring)

File: instrumentation/dalli/lib/opentelemetry/instrumentation/dalli/utils.rb
Comment (fbogsany) [reply instrumentation/dalli/lib/opentelemetry/instrumentation/dalli/utils.rb:43]: It's a pretty trivial boolean check - I think this is fine.

File: instrumentation/dalli/lib/opentelemetry/instrumentation/dalli/utils.rb
Comment (fbogsany) [inline instrumentation/dalli/lib/opentelemetry/instrumentation/dalli/utils.rb:50]: Do you need to prepend with `Utils.` here? We're in the same module.

File: instrumentation/dalli/lib/opentelemetry/instrumentation/dalli/utils.rb
Comment (fbogsany) [inline instrumentation/dalli/lib/opentelemetry/instrumentation/dalli/utils.rb:61]: Why the hash options here instead of keyword args with defaults?

File: instrumentation/dalli/lib/opentelemetry/instrumentation/dalli/utils.rb
Comment (fbogsany) [inline instrumentation/dalli/lib/opentelemetry/instrumentation/dalli/utils.rb:57]: I really wish we could do this truncation while formatting. I looked for options to do that, but didn't find anything. My concern is that we could have a multi-megabyte binary blob that we're going to go to the trouble of encoding, only to throw away or truncate the result.

File: instrumentation/dalli/test/opentelemetry/instrumentation/dalli/instrumentation_test.rb
Comment (fbogsany) [reply instrumentation/dalli/test/opentelemetry/instrumentation/dalli/instrumentation_test.rb:76]: Is there an equivalent of `Dalli::Server.any.stub(...)` to avoid reaching in to the instance variable?

Otherwise this seems fine to me. You could go to the trouble of pulling in Semian, but I'm not sure it's worth it yet.

Comment (fbogsany) [reply instrumentation/dalli/test/opentelemetry/instrumentation/dalli/instrumentation_test.rb:76]: Hmm, doesn't look like Semian would help here - it doesn't include support for Dalli AFAICT.

Comment (robertlaurin) [reply instrumentation/dalli/test/opentelemetry/instrumentation/dalli/instrumentation_test.rb:76]: >Is there an equivalent of Dalli::Server.any.stub(...) to avoid reaching in to the instance variable?

I've been looking, and it seems minitest doesn't have this baked in and I didn't really want to pull in another test dependency for just this one test.

File: instrumentation/dalli/lib/opentelemetry/instrumentation/dalli/utils.rb
Comment (robertlaurin) [reply instrumentation/dalli/lib/opentelemetry/instrumentation/dalli/utils.rb]: Nope, removed.

File: instrumentation/dalli/lib/opentelemetry/instrumentation/dalli/utils.rb
Comment (robertlaurin) [reply instrumentation/dalli/lib/opentelemetry/instrumentation/dalli/utils.rb]: I just copied it over from DD 🙈 , fixed.

File: instrumentation/dalli/lib/opentelemetry/instrumentation/dalli/utils.rb
Comment (fbogsany) [reply instrumentation/dalli/lib/opentelemetry/instrumentation/dalli/utils.rb:39]: I'm kinda OK with handling this case later. As it is, the default in `opname` will catch this and turn it into `"sasl_authentication"` in both cases.

**Review by fbogsany — APPROVED** [review]
LGTM after the merge conflicts are fixed. Thanks @robertlaurin!

File: instrumentation/dalli/lib/opentelemetry/instrumentation/dalli/utils.rb
Comment (robertlaurin) [reply instrumentation/dalli/lib/opentelemetry/instrumentation/dalli/utils.rb:39]: Yeah trying to dig into the `sasl_authentication` method here https://github.com/petergoldstein/dalli/blob/master/lib/dalli/server.rb#L701-L730 and I'm not sure how I could extract more meaningful info out of it right now.  This puts us in a better place to iterate on it going forward when/if needed.

Comment (ericmustin) [reply instrumentation/dalli/lib/opentelemetry/instrumentation/dalli/utils.rb:39]: works for me 👍

**Review by fbogsany — APPROVED** [review]

**Comment by fbogsany** [comment]
There's a conflict in the Rakefile and this gem should be added to the `releases.yml` file. Otherwise LGTM, thanks!

**Pull request merged**
